### ISSUE TITLE
CallbackOnCollectedDelegate when leaving my app

### ISSUE THREAD

**Issue by OoDeLally** [ZeBobo5/Vlc.DotNet#660]
# Generic information

- **Vlc.DotNet version** : 3.1.0
- **Vlc.DotNet project used** : WPF with WinForms
- **libvlc version** : Nuget package VideoLAN.LibVLC Windows 3.0.11
- **.net version** : 4.7.2
- **Project language** : C#
- **Project build architecture** : AnyCPU
- **Operating system** : Windows 10 Pro

# Summary

I am able to read a video. However when leaving my app, I get the following exception:
```
Managed Debugging Assistant 'CallbackOnCollectedDelegate' 
  Message=Managed Debugging Assistant 'CallbackOnCollectedDelegate' : 
    'A callback was made on a garbage collected delegate of type 'Vlc.DotNet.Core.Interops!Vlc.DotNet.Core.Interops.Signatures.EventCallback::Invoke'. 
    This may cause application crashes, corruption and data loss. When passing delegates to unmanaged code, they must be kept alive by the managed application until it is guaranteed that they will never be called.'
```

I've read this thread https://github.com/ZeBobo5/Vlc.DotNet/issues/397

From what I understand, the managed part GCed the callback it has previously given to the unmanaged part.

Is the solution to prevent GCing these callbacks? If so I'm not sure which callback we are talking about. I don't create any callback, I use a rather vanilla minimal setup. This seems to happen internally, and I feel pretty powerless.

Here is my setup:

```xaml
<!-- MainWindows.xaml -->
<WindowsFormsHost x:Name="WindowsFormsHost" />
```

```cs
public partial class MainWindow : Window
    {
        private VlcControl vlcControl;


        public void InitPlayer()
        {
            vlcControl = new VlcControl();
            WindowsFormsHost.Child = vlcControl;
            vlcMediaPlayer = vlcControl.VlcMediaPlayer;

            var currentAssembly = Assembly.GetEntryAssembly();
            var currentDirectory = new FileInfo(currentAssembly.Location).DirectoryName;
            // Default installation path of VideoLAN.LibVLC.Windows
            var libDirectory = new DirectoryInfo(Path.Combine(currentDirectory, "libvlc", IntPtr.Size == 4 ? "win-x86" : "win-x64"));

            vlcControl.BeginInit();
            vlcControl.VlcMediaplayerOptions = new string[] {
                "--file-logging", "-vvv", "--logfile=vlc-control.log"
            };
            vlcControl.VlcLibDirectory = libDirectory;
            vlcControl.EndInit();

            // Once VLC is asked to play something, then it crashes upon closing.
            // If we dont play anything, it closes gracefully.
            vlcControl.Play(new Uri("D:\\movie.avi"));
        }
}
```

Am I doing something wrong?

Here is what I tried so far:

- Declare `WindowsFormsHost`, `VlcControl` and `VlcMediaPlayer` as `static` so it doesn't get GCed prematurely.

-  `ResetMedia` on windows closing:
```cs
        private void OnClosed(object sender, EventArgs e)
        {
            if (vlcControl != null)
            {
                vlcControl.ResetMedia();
            }
        }
```

- Reassign the callbacks as suggested in https://github.com/ZeBobo5/Vlc.DotNet/issues/397#issuecomment-375575626
But I had to put `null` everywhere, since I really don't know what to do in these callback... And it crashes when I use `IntPtr.Zero` as `userData`. Not sure what I should use...

Thanks a lot!

- [X ] I confirm that my issue doesn't happen in VLC itself.

**Comment by jeremyVignelles** [comment]
Are you listening to any event? I fail to see which callback it's complaining about.
CallbackOnCollectedDelegate often indicates something has gone horribly wrong in the object lifetime management and that should be fixed.

EDIT : I see that you original message listened to EncounteredError. Does it crash with the code you've sent here?

**Comment by OoDeLally** [comment]
> Are you listening to any event? I fail to see which callback it's complaining about.

Thank you for asking that. While searching to make sure I'm not listening to anything, I realized that I was calling my `InitPlayer()` function twice. Once in `MainWindow()`, and once in `MainWindow.OnLoaded()`.
This must lead the exact same situation as https://github.com/ZeBobo5/Vlc.DotNet/issues/397 where OP was trying to have multiple instances.

Since I removed the duplicate, it now closes gracefully.

I spend hours on this... Thanks for making me search around :D

 
> EDIT : I see that you original message listened to EncounteredError. Does it crash with the code you've sent here?

Yes, I removed it to keep it simple, since having it or not has no impact on the outcome.
